### PR TITLE
Add a WafPackage base class

### PR DESCRIPTION
**Pull request by adamjstewart**

### CODE REVIEW

File: var/spack/repos/builtin/packages/tut/package.py [hunk 1]
```diff
@@ -25,17 +25,19 @@
 from spack import *
 
 
-class Tut(Package):
+class Tut(WafPackage):
```
Comment (junghans) [inline var/spack/repos/builtin/packages/tut/package.py]: Nice!

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/tut/package.py]: Yeah, I got bored. Adding new build systems is kind of fun!

Comment (alalazo) [inline var/spack/repos/builtin/packages/tut/package.py]: And you didn't try how fun it is doing the machinery with metaclasses :laughing:

**Comment by adamjstewart** [comment]
It's actually a very interesting little build system. Check out the documentation if you can. They go to great lengths to ensure that builds work efficiently in parallel.

**Comment by adamjstewart** [comment]
@citibeth `waf` is a Python script.

**Comment by citibeth** [comment]
Yes... and so is Spack.  So.. .can Spack run it without launching a new
Python interpreter?  Or at least, query the current Python interpreter for
its path, and then run that in a new process.  No need to install a new
Python.

**Comment by citibeth** [comment]
Just use `sys.exectuable`.  See:

http://stackoverflow.com/questions/2589711/find-full-path-of-the-python-interpreter

**Comment by adamjstewart** [comment]
It sounds like this is a question of whether or not _any_ package should have a Python build dependency. Up until now, we've said "yes", if a package needs Python to build, it should explicitly declare that dependency. This allows you to constrain the versions of Python that can be used. Now that Spack supports both Python 2 and 3, this will be even more important, as many packages cannot be built with Python 3. I'm not sure which versions of Python Waf supports, but I'd like to play it on the safe side and explicitly declare the dependency.

**Comment by tgamblin** [comment]
If you unconditionally use the Spack python in a build that requires Python 2/3, and Spack is version 3/2, I believe this will break because spack's `python` will be unable to run properly in the build environment.

@adamjstewart: That said, I think you *could* actually pull off what @citibeth is suggesting.  If the default `build_environment.py` stuck a `python` `Executable` into the module scope by default, and if the `PythonPackage` overrode that with its own, then you wouldn't need `WafPackage` to implicitly depend on `Python`.  It would always have a Python that could run.  I like the idea of minimizing unnecessary builds.  Then again, both the WAF packages used so far *do* depend on Python.  I suspect a lot of `WafPackage`s will.

**Comment by jcftang** [comment]
waf should work quite well with python 2/3 and you should probably depend on the system version of python and not build your own just to run it.  Other than that +1

**Comment by adamjstewart** [comment]
I'll open up a separate issue to discuss this. This isn't restricted to just `WafPackage`, but could be used for any package that currently has a build dependency on Python. I don't want this to hold up this PR.

**Pull request merged**